vpr/src/route/connection_box_lookahead_map.cpp

```diff
@@ -21,370 +22,386 @@
+    if (src.delay < dst->delay) {
+        dst->delay = src.delay;
+        dst->congestion = src.congestion;
     }
+}
 
-    /* run Dijkstra's algorithm for each segment type & channel type combination */
-    for (int iseg = 0; iseg < (ssize_t)num_segments; iseg++) {
-        VTR_LOG("Creating cost map for %s(%d)\n",
-                segment_inf[iseg].name.c_str(), iseg);
-        /* allocate the cost map for this iseg/chan_type */
-        t_routing_cost_map cost_map;
-
-        int count = 0;
-        for (const auto loc : search_locations) {
-            count += search_at(iseg, loc.first, loc.second, &cost_map);
-        }
-
-        if (count == 0) {
-            VTR_LOG_WARN("Segment %s(%d) found no start_node_ind\n",
-                         segment_inf[iseg].name.c_str(), iseg);
+// find the minimum cost entry from the nearest manhattan distance neighbor
+std::pair<Cost_Entry, int> CostMap::get_nearby_cost_entry(const vtr::NdMatrix<Cost_Entry, 2>& matrix,
+                                                          int cx,
+                                                          int cy,
+                                                          const vtr::Rect<int>& bounds) {
+    // spiral around (cx, cy) looking for a nearby entry
+    int n = 1;
+    bool in_bounds;
+    Cost_Entry entry;
+
+    do {
+        in_bounds = false;
+        for (int ox = -n; ox <= n; ox++) {
+            int x = cx + ox;
+            int oy = n - abs(ox);
+            int yp = cy + oy;
+            int yn = cy - oy;
+            if (bounds.contains(vtr::Point<int>(x, yp))) {
+                assign_min_entry(&entry, matrix[x][yp]);
+                in_bounds = true;
+            }
+            if (bounds.contains(vtr::Point<int>(x, yn))) {
+                assign_min_entry(&entry, matrix[x][yn]);
+                in_bounds = true;
+            }
         }
-
-#if 0
-        for(const auto & e : cost_map) {
-            VTR_LOG("%d -> %d (%d, %d): %g, %g\n",
-                    std::get<0>(e).first, std::get<0>(e).second,
-                    std::get<1>(e).first, std::get<1>(e).second,
-                    std::get<2>(e).delay, std::get<2>(e).congestion);
+        if (entry.valid()) return std::make_pair(entry, n);
+        n++;
+#if defined(FILL_LIMIT)
+        if (n > FILL_LIMIT) {
+            break;
         }
 #endif
+    } while (in_bounds);
+    return std::make_pair(Cost_Entry(), n);
+}
 
-        /* boil down the cost list in routing_cost_map at each coordinate to a
-         * representative cost entry and store it in the lookahead cost map */
-        g_cost_map.set_cost_map(iseg, cost_map,
-                                REPRESENTATIVE_ENTRY_METHOD);
-    }
+// sample on a uniformly spaced grid within a rectangle
+// sample(vtr::Rect(N), 0, 0, M) is vtr::Point(0, 0)
+// sample(vtr::Rect(N), M, M, M) is vtr::Point(N, N)
+// To avoid the edges, use `sample(r, x+1, y+1, N+1) for x, y, in 0..N-1
+template<typename T>
```

Comment:
Probably worth making a couple unit tests for this function.

Reply:
Done. I moved `sample()` to `vtr_geometry.hpp` and added a test.